tasking/pulp/tasking/base.py

```diff
@@ -0,0 +1,444 @@
+        tag_list = kwargs.get('tags', [])
+        group_id = kwargs.get('group_id', None)
+
+        # Set the parent attribute if being dispatched inside of a Task
+        parent_arg = self._get_parent_arg
+
+        # Create a new task status with the task id and tags.
+        TaskStatus.objects.create(pk=inner_task_id, state=TaskStatus.WAITING, tags=tag_list,
+                                  group=group_id, **parent_arg)
+
+        # Call the outer task which is a promise to call the real task when it can.
+        _queue_reserved_task.apply_async(args=[task_name, inner_task_id, resource_id, args, kwargs],
+                                         queue=RESOURCE_MANAGER_QUEUE)
+        return AsyncResult(inner_task_id)
+
+    def apply_async(self, *args, **kwargs):
+        """
+        A wrapper around the super() apply_async method. It allows us to accept a few more
+        arguments than Celery does for our own purposes, listed below. It also allows us
+        to create and update task status which can be used to track status of this task
+        during it's lifetime.
+
+        :param queue:       The queue that the task has been placed into (optional, defaults to
+                            the general Celery queue named 'celery'.)
+        :type  queue:       basestring
+
+        :param tags:        A list of tags (strings) to place onto the task, used for searching for
+                            tasks by tag
+        :type  tags:        list
+
+        :param group_id:    The id that identifies which group of tasks a task belongs to
+        :type group_id:     uuid.UUID
+
+        :return:            An AsyncResult instance as returned by Celery's apply_async
+        :rtype:             celery.result.AsyncResult
+        """
+        tag_list = kwargs.pop('tags', [])
+        group_id = kwargs.pop('group_id', None)
+        async_result = super(UserFacingTask, self).apply_async(*args, **kwargs)
+        async_result.tags = tag_list
+
+        # Set the parent attribute if being dispatched inside of a Task
+        parent_arg = self._get_parent_arg
+
+        # Create a new task status with the task id and tags.
+        TaskStatus.objects.create(pk=async_result.id, state=TaskStatus.WAITING, tags=tag_list,
+                                  group=group_id, **parent_arg)
+        return async_result
+
+    def __call__(self, *args, **kwargs):
+        """
+        Update the :class: `pulp.app.models.Task` object, log, and call the task.
+
+        This updates the :attr: `started_at` and sets the :attr: `state` to :attr: `RUNNING`.
+
+        If the :class: `pulp.app.models.Task` status attribute :attr: `state` is
+        :attr: `CANCELED` the task is skipped.
+
+        Skip the status updating if the task is called synchronously.
+        """
+        if not self.request.called_directly:
+            task_status = TaskStatus.objects.get(pk=self.request.id)
+            if task_status.state != TaskStatus.WAITING:
```

Comment:
I'd consider moving the rest of this block into a method on the TaskStatus model. That would allow it to manage its own state change and any consistency rules among attributes. As-is, this code expresses and enforces consistency rules (1. state should be WAITING when moving to RUNNING, 2. ``started_at`` value must be set when moving to RUNNING) far from the actual data structure.

Instead, you'd just call something like:

``task_status.set_running()``